### DISCUSSION TITLE
Low FPS even though CPU or GPU appears to have low utilization

### DISCUSSION THREAD

**Discussion by Karsten75** [KnuckleCracker/CW4-bug-tracker#1027] [General]
From a Steam Discussion thread: 
 _especially at higher speed, on certain user created map, im getting 20 to 30 fps even tho gpu usage and cpu usage only hovering around 30%, anyway to get higher fps by forcing game to use more power?_
 
 Answer:
 It does not work that way.

 -  GPU utilization
 the game is not graphics intensive for high-end GPUs and should never max out a huger-end GPU.

  - CPU utilization:
There is no code in the game that deliberately looks to cap CPU consumption. This assumption is erroneous.

There are two possible causes here.

1. It is not feasible to multi-task all aspects of the game. What was originally considered CPU-intensive tasks, such as the Creeper slow sim is multi-tasked. That still leaves other parts of the game as single-threaded - I suspect many games are like this, because there are costs in CPU as well as complexity to multi-task all components of the game.

In this category are custom units, and things like that. If you have large numbers of units, and they all move, it chews up a single thread of CPU. Each unit does not consume much, but together.

2. Some custom maps have huge amounts of scripting and this consumes CPU on a single thread.

If neither of the above applies, then there is some other bottleneck, that does not show up in your monitoring tools. I can't hazard a guess as to what it may be, but I can assure you that the game has no code to attempt to limit CPU consumption as you claim.

**Comment by knucracker** [comment]
When noting CPU usage using the windows Task manager, right click the CPU graph and select "Change Graph To: Logical Processors".  That will show a breakdown of each core and it's usage.  If your CPU has multiple cores (and most do), overall usage might appear to be X% even when one core is near 100%.  This happens because the other cores are idle.  Showing the breakdown per core/logical processor will be more informative.

As noted, custom game logic in a map is all single threaded and will run on one core.  So a map with heavy user scripting (4rpl) will work one core very heavily.
